docs/spec.md

```diff
@@ -173,10 +188,48 @@ values of type `tensor<T>`).
   and an **imaginary part** of the same **element type**. Supported complex
   types are `complex<f32>` (both parts are of type `f32`) and `complex<f64>`
   (both parts are of type `f64`).
-* In the future, we are also planning to introduce **quantized types** that
-  represent integer values obtained via uniform quantization of floating-point
-  values using given scales and zero points
-  ([#588](https://github.com/openxla/stablehlo/issues/588)).
+* **Quantized types** represent integer values of a **storage type** in the
+  range from `storage_min` to `storage_max` (inclusive) that correspond to
+  floating-point values of an **expressed type**. For a given integer value `i`,
+  the corresponding floating-point value `f` can be computed as
+  `f = (i - zero_point) * scale`, where `scale` and `zero_point` are called
+  **quantization parameters**.
+
+Quantization can be **per-tensor**, meaning, having one `scale` and/or
+`zero_point` for the entire tensor or can be **per-axis**, meaning, having
+multiple `scales` and/or `zero_points`, one pair per slice of a particular
+dimension `quantized_dimension`. More formally, in a tensor `t` of per-axis
+quantized element type, there are `dim(t, quantized_dimension)` slices of the
+`quantized_dimension`: `t[:, ..., 0, ..., :], t[:, ..., 1, ..., :]`, etc. All
+elements in the `i`th slice use `scales[i]` and `zero_points[i]` as their
+quantization parameters.
+
+Quantized types satisfy the following constraints:
+
+* (C1) `num_bits(storage_type) < num_bits(expressed_type)`.
+* (C2) `type(storage_min) = storage_type`.
+* (C3) `type(storage_max) = storage_type`.
+* (C4) `storage_min` and `storage_max` have the following values:
+  * If `is_signed(storage_type)` and `d = num_bits(storage_type)`:
+    * `storage_min = -2^(d-1)`, `storage_max = 2^(d-1)-1` OR
+    * `storage_min = -2^(d-1)+1`, `storage_max = 2^(d-1)-1`.
+  * If `is_unsigned(storage_type)`:
+    * `storage_min = 0`, `storage_max = 2^d - 1`.
```

Comment:
Are `storage_min` and `storage_max` user-specified values or derived values from the storage type?

Reply:
With the above change of grammar, the `storage_min` and `storage_min` can be optional in which case the value can be inferred using a mix of 

(1) The general constraints 
```
* (C4) `storage_min` and `storage_max` have the following values:
  * If `is_signed(storage_type)` and `d = num_bits(storage_type)`:
    * `storage_min = -2^(d-1)`, `storage_max = 2^(d-1)-1` OR
    * `storage_min = -2^(d-1)+1`, `storage_max = 2^(d-1)-1`.
  * If `is_unsigned(storage_type)`:
    * `storage_min = 0`, `storage_max = 2^d - 1`.
```

(2) Op specific constraints, like for AddOp we have
```
  * (C4) `abs(storage_min(lhs)) != abs(storage_max(lhs))`.
```

Reply:
Given that storage_min and storage_max are optional in the new grammar, where do we specify their default values?

Reply:
I agree that we should clarify that. 
This is my proposal: 
```
**Quantized types** represent ... are called  **quantization parameters**. The `storage_min` and `storage_max`, being optional, have default values which should satisfy the associated constraints
 as laid out in this document.
```

Please take a look.

Reply:
Incorporated the proposal in the specification. Feel free to review.

Reply:
Instead of:

> `storage_min` and `storage_max` have the following values:

Suggestion:

> If `storage_min` and `storage_max` are unspecified, the default values are defined as:

Reply:
Thinking about this some more, do we have to explicitly specify the default values for storage_min and storage_max? We can use the "is empty" predicate that we used for quantization_dimension below to account for the case when they are missing.

Also, are storage_min and storage_max fully customizable by the user or not? I.e. can they take just one of the 1-2 support values and this is it, or they can be arbitrary, as long as C7 is satisfied?

Reply:
Thanks for the suggestion!

>> Also, are storage_min and storage_max fully customization by the user or not? I.e. can they take just one of the 1-2 support values and this is it, or they can be arbitrary, as long as C7 is satisfied?

Per https://github.com/llvm/llvm-project/blob/215f36f7917d87f81eeeee0de8ec55d11d83a164/mlir/lib/Dialect/Quant/IR/QuantTypes.cpp#L54, they can be somewhat arbitrary as long as the the following constraints are satisfied.

```
*  If is_signed(storage_type) and d = num_bits(storage_type):
      * storage_min >= -2^(d-1), storage_max <= 2^(d-1)-1 OR
* If is_unsigned(storage_type):
      * storage_min >= 0, storage_max <= 2^d - 1.
* `storage_max - storage_min > 0`.
``` 
 
>> We can use the "is empty" predicate that we used for quantization_dimension below to account for the case when they are missing.

I agree with that.

We can specify the semantics, of missing `storage_min` and `storage_max`, using the predicate as follows ([ref](https://github.com/llvm/llvm-project/blob/215f36f7917d87f81eeeee0de8ec55d11d83a164/mlir/lib/Dialect/Quant/IR/TypeParser.cpp#L67))

```
* If `storage_min` is empty, then `storage_min = is_signed(storage_type) ? -2^(d-1) : 0`.
* If `storage_max` is empty, then `storage_max = is_signed(storage_type) ? 2^(d-1) : 2^d - 1`.
  where d = num_bits(storage_type)
```

>> do we have to explicitly specify the default values for storage_min and storage_max? 

IMO, we do not have to specify the above default values explicitly, but add them as constraints. With that, as you can imagine, we can have multiple constraints base on the predicates "is_empty" and "is_signed".
Else, we can keep the constraints separate from the specification of default values. As you can see, the constraints will be equally applicable when min/max is empty or not.

Reply:
Left a few follow-up comments on the updated text. Let's resolve this thread and proceed in individual new threads.